pkg/rbac/sharedData.go

```diff
@@ -32,116 +39,164 @@ type resource struct {
+				continue
 			}
-		}
 
-		// get all namespaces in cluster and cache in shared.namespaces.
-		shared.nsLock.Lock()
-		defer shared.nsLock.Unlock()
-		allNamespaces, nsErr := shared.GetSharedNamespaces(cache, ctx)
-		if nsErr != nil {
-			shared.nsErr = nsErr
-		}
+			shared.csResources = append(shared.csResources, resource{apigroup: apigroup, kind: kind})
 
-		// update the cache.
-		shared.namespaces = allNamespaces
-		shared.csUpdatedAt = time.Now()
-	} else {
-		klog.V(8).Info("Using cluster scoped resources from cache.")
+		}
 	}
+	shared.csUpdatedAt = time.Now()
 
-	return shared.csResources, shared.csErr
+	return shared.csErr
 }
 
-func (shared *SharedData) GetSharedNamespaces(cache *Cache, ctx context.Context) ([]string, error) {
+func (shared *SharedData) GetSharedNamespaces(cache *Cache, ctx context.Context) error {
+	shared.nsLock.Lock()
+	defer shared.nsLock.Unlock()
+	//empty previous cache
+	shared.namespaces = nil
+	shared.nsErr = nil
 
-	if len(shared.namespaces) > 0 &&
-		time.Now().Before(shared.nsUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
-		klog.V(5).Info("Using namespaces from shared cache")
+	klog.V(5).Info("Getting namespaces from Kube Client..")
 
-	} else {
+	clientset, kubeErr := kubernetes.NewForConfig(cache.restConfig)
+	if kubeErr != nil {
+		klog.Warning("Error with creating a new clientset.", kubeErr.Error())
+		shared.nsErr = kubeErr
+		return shared.nsErr
+	}
 
-		klog.V(5).Info("Getting namespaces from Kube Client..")
+	namespaceList, nsErr := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
 
-		cache.restConfig = config.GetClientConfig()
-		clientset, err := kubernetes.NewForConfig(cache.restConfig)
-		if err != nil {
-			klog.Warning("Error with creating a new clientset.", err.Error())
+	if nsErr != nil {
+		klog.Warning("Error resolving namespaces from KubeClient: ", nsErr)
+		shared.nsErr = nsErr
+		return shared.nsErr
+	}
 
-		}
+	// add namespaces to allNamespace List
+	for _, n := range namespaceList.Items {
+		shared.namespaces = append(shared.namespaces, n.Name)
+	}
+	shared.nsUpdatedAt = time.Now()
 
-		namespaceList, kubeErr := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	return shared.nsErr
+}
 
-		if kubeErr != nil {
-			klog.Warning("Error resolving namespaces from KubeClient: ", kubeErr)
-			shared.nsErr = kubeErr
-			return shared.namespaces, shared.nsErr
-		}
+func (shared *SharedData) GetSharedManagedClusters(cache *Cache, ctx context.Context) error {
 
-		// add namespaces to allNamespace List
-		for _, n := range namespaceList.Items {
-			shared.namespaces = append(shared.namespaces, n.Name)
-		}
-		shared.nsUpdatedAt = time.Now()
+	shared.mcLock.Lock()
+	defer shared.mcLock.Unlock()
+	// clear previous cache
+	shared.managedClusters = nil
+	shared.mcErr = nil
 
+	var managedClusters []string
+
+	resourceObj, err := cache.dynamicClient.Resource(clusterVersionGvr).List(context.TODO(), metav1.ListOptions{})
```

Comment:
We have a context, so let's use it.
```suggestion
	resourceObj, err := cache.dynamicClient.Resource(clusterVersionGvr).List(ctx, metav1.ListOptions{})
```